examples/fluids/src/misc.c

```diff
@@ -150,32 +136,66 @@ PetscErrorCode RegressionTests_NS(AppCtx app_ctx, Vec Q) {
   PetscFunctionReturn(PETSC_SUCCESS);
 }
 
-// Get error for problems with exact solutions
-PetscErrorCode GetError_NS(CeedData ceed_data, DM dm, User user, Vec Q, PetscScalar final_time) {
-  PetscInt  loc_nodes;
-  Vec       Q_exact, Q_exact_loc;
-  PetscReal rel_error, norm_error, norm_exact;
+PetscErrorCode ComputeL2Error(Vec Q_loc, PetscReal l2_error[5], OperatorApplyContext op_error_ctx, CeedContextFieldLabel time_label,
+                              CeedScalar time) {
+  Vec       E;
+  PetscReal l2_norm[5];
+  PetscFunctionBeginUser;
+
+  if (time_label) CeedOperatorSetContextDouble(op_error_ctx->op, time_label, &time);
+  PetscCall(VecDuplicate(Q_loc, &E));
+  PetscCall(ApplyCeedOperatorLocalToGlobal(Q_loc, E, op_error_ctx));
+  PetscCall(VecStrideNormAll(E, NORM_1, l2_norm));
+  for (int i = 0; i < 5; i++) l2_error[i] = sqrt(l2_norm[i]);
+
+  PetscCall(VecDestroy(&E));
+
+  PetscFunctionReturn(PETSC_SUCCESS);
+};
+
+// Get error for problems with true solutions
+PetscErrorCode GetError_NS(CeedData ceed_data, DM dm, User user, ProblemData *problem, Vec Q, PetscScalar final_time) {
+  Vec         Q_loc;
+  PetscReal   l2_error[5];
+  const char *state_var_source = "Conservative";
   PetscFunctionBegin;
 
-  // Get exact solution at final time
-  PetscCall(DMCreateGlobalVector(dm, &Q_exact));
-  PetscCall(DMGetLocalVector(dm, &Q_exact_loc));
-  PetscCall(VecGetSize(Q_exact_loc, &loc_nodes));
-  PetscCall(ICs_FixMultiplicity(dm, ceed_data, user, Q_exact_loc, Q_exact, final_time));
+  // Get the local values of the final solution
+  PetscCall(DMGetLocalVector(dm, &Q_loc));
+  PetscCall(DMGlobalToLocal(dm, Q, INSERT_VALUES, Q_loc));
+  PetscCall(DMPlexInsertBoundaryValues_NS(dm, PETSC_TRUE, Q_loc, final_time, NULL, NULL, NULL));
+
+  // Compute the L2 error in the source state variables
+  if (user->phys->ics_time_label) CeedOperatorSetContextDouble(ceed_data->op_ics_ctx->op, user->phys->ics_time_label, &final_time);
```

Comment:
Prefer using `UpdateContextLabel`, but that's currently not in `navierstokes.h`; feel free to add the signature there are use it here if you want.